src/NUnitFramework/framework/Internal/DefaultBlockingAwaitAdapter.cs

```diff
@@ -0,0 +1,69 @@
+// ***********************************************************************
+// Copyright (c) 2018 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System.Threading;
+
+namespace NUnit.Framework.Internal
+{
+    /// <summary>
+    /// Useful when wrapping awaitables whose <c>GetResult</c> method does not block until complete.
+    /// Contains a default mechanism to implement <see cref="AwaitAdapter.BlockUntilCompleted"/>
+    /// via <see cref="AwaitAdapter.IsCompleted"/> and <see cref="AwaitAdapter.OnCompleted"/>.
+    /// </summary>
+    internal abstract class DefaultBlockingAwaitAdapter : AwaitAdapter
+    {
+        private volatile ManualResetEventSlim _completedEvent;
+
+        public sealed override void BlockUntilCompleted()
+        {
+            if (IsCompleted) return;
+
+            var completedEvent = _completedEvent; // Volatile read (would be Volatile.Read if not for net40 support)
```

Comment:
Question (rather than a suggestion), but why not wrap this in a `#if NET_40` and then use `Volatile.Read()` in an `#else` block?

Reply:
Using `Volatile.Read` would be redundant so long as `_completedEvent` is declared with the `volatile` keyword, and in fact, the compiler would warn that a `volatile` field is unsafely used as a `ref` parameter. So I could do that, but I would also have to add `#if` around the `volatile` keyword and I didn't feel like the effort was worth it since the behavior is the same.

Reply:
Oh ok, that makes sense. Thanks for the explanation!

Reply:
I forgot: recent versions of the C# compiler suppress the warning for `ref` arguments to Interlocked and Volatile methods.